src/Pact/Runtime/Typecheck.hs

```diff
@@ -0,0 +1,177 @@
+-- the pair is the type variable itself and the term type.
+typecheckTerm :: forall e . Info -> Type (Term Name) -> Term Name
+       -> Eval e (Maybe (TypeVar (Term Name),Type (Term Name)))
+typecheckTerm i spec t = do
+
+  ty <- case typeof t of
+    Left s -> evalError i $ "Invalid type in value location: " <> pretty s
+    Right r -> return r
+
+  let
+
+    tcFail found = typecheckFailed i found spec
+
+    tcOK = return Nothing
+
+    -- | check container parameterized type.
+    -- 'paramCheck pspec pty check' check specified param ty 'pspec' with
+    -- value param ty 'pty'. If not trivially equal, use 'check'
+    -- to determine actual container value type, and compare for equality
+    -- with specified.
+    paramCheck :: Type (Term Name)
+               -> Type (Term Name)
+               -> (Type (Term Name) -> Eval e (Type (Term Name)))
+               -> Eval e (Maybe (TypeVar (Term Name),Type (Term Name)))
+    paramCheck TyAny _ _ = tcOK -- no spec
+    paramCheck pspec pty check
+      | pspec == pty = tcOK -- equality OK
+      | otherwise = do
+          -- run check function to get actual content type
+          checked <- check pspec
+          -- final check expects full match with toplevel 'spec'
+          if checked == spec then tcOK else tcFail checked
+
+    -- | infer list value type
+    checkList es lty = return $ TyList $
+                    case nub (map typeof $ V.toList es) of
+                      [Right a] -> a -- uniform value type: return it
+                      [] -> lty -- empty: return specified
+                      _ -> TyAny -- otherwise untyped
+
+  case (spec,ty,t) of
+    (_,_,_) | spec == ty -> tcOK -- identical types always OK
+    (TyAny,_,_) -> tcOK -- var args are untyped
+    (TyVar {..},_,_) ->
+      if spec `canUnifyWith` ty
+      then return $ Just (_tyVar,ty) -- collect found types under vars
+      else tcFail ty -- constraint failed
+    -- check list
+    (TyList lspec,TyList lty,TList {..}) ->
+      paramCheck lspec lty (checkList _tList)
+    -- check object
+    (TySchema TyObject ospec specPartial,TySchema TyObject oty _,TObject {..}) ->
+      paramCheck ospec oty (checkUserType specPartial i (_oObject _tObject))
+    (TyPrim (TyGuard a),TyPrim (TyGuard b),_) -> case (a,b) of
+      (Nothing,Just _) -> tcOK
+      (Just _,Nothing) -> tcOK
+      (c,d) -> if c == d then tcOK else tcFail ty
+    _ -> tcFail ty
+
+-- | check object args. Used in 'typecheckTerm' above and also in DB writes.
+-- Total flag allows for partial row types if False.
+checkUserType :: SchemaPartial -> Info -> ObjectMap (Term Name) -> Type (Term Name) -> Eval e (Type (Term Name))
+checkUserType partial i (ObjectMap ps) (TyUser tu@TSchema {..}) = do
+  -- fields is lookup from name to arg.
+  -- TODO consider OMap or equivalent for schema fields
+  let fields = M.fromList . map (FieldKey . _aName &&& id) $ _tFields
+  aps <- forM (M.toList ps) $ \(k,v) -> case M.lookup k fields of
```

Comment:
`M.traverseWithKey` avoids some O(n) conversions here and below

Reply:
Let's see ... now it's
here: M.toList (n), forM (n) 
'findMissing`: M.fromList (n*log n), map (n) 
`typecheck`: no change
Total: (n*log n) + 3n

Whereas:
here with `traverseWithKey`: n
`findMissing`: `mapKeys` (n*log n). NB: We can't use unsafe `mapKeysMonotonic` O(n) because `Arg` is too different from `FieldKey`, at least without validating that the names in the `Arg` list are unique, which they should be, but a corrupted map is not something I want to contemplate.
`typecheck`: M.toList (n)
Total: (n*log n) + 2n

Small savings, but why not, runtime typechecking is a pain point in performance. Maybe take another look at this stuff to see if there are other improvements.